security-oauth2/src/main/java/io/micronaut/security/oauth2/routes/OauthRouteBuilder.java

```diff
@@ -106,17 +108,21 @@
                         LOG.debug("Registering callback route [GET: {}] for oauth configuration [{}]", callbackPath, name);
                         LOG.debug("Registering callback route [POST: {}] for oauth configuration [{}]", callbackPath, name);
                     }
-                    buildRoute(HttpMethod.GET, callbackPath, executionHandle);
-                    buildRoute(HttpMethod.POST, callbackPath, executionHandle).consumes(MediaType.APPLICATION_FORM_URLENCODED_TYPE);
+                    buildRoute(HttpMethod.GET, callbackPath, executionHandle).produces(MediaType.TEXT_HTML_TYPE);
+                    buildRoute(HttpMethod.POST, callbackPath, executionHandle)
+                            .produces(MediaType.TEXT_HTML_TYPE)
+                            .consumes(MediaType.APPLICATION_FORM_URLENCODED_TYPE);
 
                     if (isDefaultProvider) {
                         final String defaultCallbackPath = oauthRouteUrlBuilder.buildCallbackUri(null).getPath();
                         if (LOG.isDebugEnabled()) {
                             LOG.debug("Registering default callback route [GET: {}] for oauth configuration [{}]", defaultCallbackPath, name);
                             LOG.debug("Registering default callback route [POST: {}] for oauth configuration [{}]", defaultCallbackPath, name);
                         }
-                        buildRoute(HttpMethod.GET, defaultCallbackPath, executionHandle);
-                        buildRoute(HttpMethod.POST, defaultCallbackPath, executionHandle).consumes(MediaType.APPLICATION_FORM_URLENCODED_TYPE);
+                        buildRoute(HttpMethod.GET, defaultCallbackPath, executionHandle).produces(MediaType.TEXT_HTML_TYPE);
+                        buildRoute(HttpMethod.POST, defaultCallbackPath, executionHandle)
```

Comment:
I think we should register two routes one for content-type `application/json` and one for application `application/x-www-form-urlencoded`